File: .claude/hooks/SessionEndSummary.ts
#!/usr/bin/env bun
/**
 * SessionEnd hook: write a final session summary to the vault's daily log.
 *
 * Fires when Claude Code exits (manual /exit, timeout, window close).
 * Reads the full session transcript from jsonl and appends a structured
 * block to <vault>/daily/YYYY-MM-DD.md so the session is recoverable
 * next time.
 *
 * Writes to <vault>/daily/YYYY-MM-DD.md — auto-detects the vault
 * directory the same way LoadMasterPrompt does (any folder named
 * `vault` or ending in `-vault`).
 *
 * Deterministic — no LLM call.
 */

import {
  findTranscript,
  readTranscript,
  formatSummaryMarkdown,
  appendToDailyLog,
  findVaultDir,
} from './lib/transcript';
import { join } from 'path';

interface HookInput {
  session_id?: string;
  cwd?: string;
}

function hhmm(ts: string | null): string {
  if (!ts) return '??:??';
  const d = new Date(ts);
  return `${String(d.getHours()).padStart(2, '0')}:${String(d.getMinutes()).padStart(2, '0')}`;
}

async function main() {
  try {
    const input = await Bun.stdin.text();
    if (!input?.trim()) return;
    const data: HookInput = JSON.parse(input);
    const sessionId = data.session_id;
    if (!sessionId) return;

    const cwd = data.cwd || process.cwd();
    const vaultDir = findVaultDir(cwd);
    if (!vaultDir) {
      console.error(
        '[SessionEndSummary] No vault directory found (looking for `vault` or `*-vault`). Skipping.',
      );
      return;
    }

    const jsonl = findTranscript(sessionId, cwd);
    if (!jsonl) return;

    const summary = readTranscript(jsonl);
    if (summary.userPrompts.length === 0) return;

    const today = new Date().toISOString().slice(0, 10);
    const dailyPath = join(vaultDir, 'daily', `${today}.md`);
    const heading = `### Session ${sessionId.slice(0, 8)} (${hhmm(summary.startTime)}–${hhmm(summary.endTime)})`;
    appendToDailyLog(dailyPath, formatSummaryMarkdown(summary, heading));
  } catch (err) {
    console.error(`[SessionEndSummary] ${err}`);
  }
}

main();

File: .claude/hooks/SessionEndSummary.py
#!/usr/bin/env python3
"""SessionEnd hook: write a final session summary to the vault's daily log.

Python equivalent of SessionEndSummary.ts. Same behaviour, same output.

Fires when Claude Code exits (manual /exit, timeout, window close).
Reads the full session transcript from jsonl and appends a structured
block to <vault>/daily/YYYY-MM-DD.md so the session is recoverable
next time.

Writes to <vault>/daily/YYYY-MM-DD.md - auto-detects the vault
directory the same way LoadMasterPrompt does (any folder named
`vault` or ending in `-vault`).

Deterministic - no LLM call.

To use this variant instead of the TypeScript one: edit
.claude/settings.json and change the SessionEnd hook command to
`python3 $CLAUDE_PROJECT_DIR/.claude/hooks/SessionEndSummary.py`.
"""

from __future__ import annotations

import json
import os
import sys
from datetime import date, datetime
from pathlib import Path

# Make the sibling `lib/` importable when this file is invoked directly.
sys.path.insert(0, str(Path(__file__).resolve().parent))
from lib.transcript import (  # noqa: E402
    append_to_daily_log,
    find_transcript,
    find_vault_dir,
    format_summary_markdown,
    read_transcript,
)


def hhmm(ts: str | None) -> str:
    if not ts:
        return "??:??"
    try:
        d = datetime.fromisoformat(ts.replace("Z", "+00:00"))
    except ValueError:
        return "??:??"
    return f"{d.hour:02d}:{d.minute:02d}"


def main() -> None:
    try:
        raw = sys.stdin.read()
        if not raw.strip():
            return
        data = json.loads(raw)
    except (json.JSONDecodeError, OSError):
        return

    session_id = data.get("session_id")
    if not session_id:
        return

    cwd = Path(data.get("cwd") or os.getcwd())
    vault_dir = find_vault_dir(cwd)
    if vault_dir is None:
        print(
            "[SessionEndSummary] No vault directory found (looking for `vault` or `*-vault`). Skipping.",
            file=sys.stderr,
        )
        return

    jsonl = find_transcript(session_id, cwd)
    if jsonl is None:
        return

    summary = read_transcript(jsonl)
    if not summary.user_prompts:
        return

    today = date.today().isoformat()
    daily_path = vault_dir / "daily" / f"{today}.md"
    short_id = session_id[:8]
    heading = (
        f"### Session {short_id} "
        f"({hhmm(summary.start_time)}–{hhmm(summary.end_time)})"
    )
    append_to_daily_log(daily_path, format_summary_markdown(summary, heading))


if __name__ == "__main__":
    try:
        main()
    except Exception as err:  # pragma: no cover - last-ditch safety net
        print(f"[SessionEndSummary] {err}", file=sys.stderr)
